### DISCUSSION TITLE
Possibility to add more 'new face' clusters? (Face recognition)

### DISCUSSION THREAD

**Discussion by sdudzin** [photoprism/photoprism#1627] [Q&A]
Is there a way to add more face clusters under `new faces` section? At first I had about 100 new faces  detected, but once those were processed (added to `recognized` tab) -  no more new face clusters are detected, even though I have a lot more faces to detect (even with similar photos).
If I just add new faces from `all faces` view then the photos are added mostly one by one which isn't practical for over 30k photos.

**Comment by rickysarraf** [comment]
I must say I'm seeing the same behavior. The initial developer builds were very good at face detection. I recollect having all faces well detected and recognized. But the final release has not had the same performance. Almost 70% of the faces on my kid are not recognized at all, any more. And in the initial developer builds, they were recognized quite well.

**Reply by lastzero** [reply]
We actually haven't changed all that much. What has been added with a recent update is that faces of babies are not automatically clustered because of #1587. If you now say it's working perfectly, we are happy to revert that. Of course it would be best to see actual images so that we don't need to rely on guessing for improvements.

When clusters have been initially identified,  new clusters shouldn't pop up - unless there actually are new faces from people that you had no pictures before.

Note that comparing faces with clusters causes high CPU load, so this isn't done in real time but in a background worker that only runs from time to time.

Last but not least you can of course change the config options to make clustering more aggressive e.g. start a cluster with less faces or when they are less similar. We figured aggressive matching creates a mess in the long term, although short-term it looks like it's really helpful.

Check out the faces CLI commands. They help with testing and you can also force updates, so that you don't have to wait for the background worker.

**Reply by lastzero** [reply]
I've updated https://docs.photoprism.org/user-guide/organize/people/ today to explain how the new faces page works and why faces might not have been detected.

**Reply by rickysarraf** [reply]
Thanks for the updated documentation. I had learnt it the hard way how important the indexed thumbnails were in facial detection. I initially had set to **Dynamic Previews** to cut down on the overall cache size. But that had a negative impact on facial detection/recognition. As a user, we need to make a balanced call on the settings.

I wonder if there should be some more documentation on the knobs you've presented in the yaml file. In particular:

```
PHOTOPRISM_FACE_SIZE | --face-size PIXELS | minimum face size in PIXELS (default: 50)
-- | -- | --
PHOTOPRISM_FACE_SCORE | --face-score QUALITY | minimum face QUALITY score (default: 9)
PHOTOPRISM_FACE_OVERLAP | --face-overlap PERCENT | face area overlap threshold in PERCENT (default: 42)
PHOTOPRISM_FACE_CLUSTER_SIZE | --face-cluster-size PIXELS | minimum size of automatically clustered faces in PIXELS (default: 80)
PHOTOPRISM_FACE_CLUSTER_SCORE | --face-cluster-score QUALITY | minimum QUALITY score of automatically clustered faces (default: 15)
PHOTOPRISM_FACE_CLUSTER_CORE | --face-cluster-core NUMBER | NUMBER of faces forming a cluster core (default: 4)
PHOTOPRISM_FACE_CLUSTER_DIST | --face-cluster-dist DISTANCE | similarity DISTANCE of faces forming a cluster core (default: 0.64)
PHOTOPRISM_FACE_MATCH_DIST | --face-match-dist OFFSET | similarity OFFSET for matching faces with existing clusters (default: 0.46)
```

**Reply by lastzero** [reply]
It's the best we can provide right now. These are standard parameters people with a background in math / machine learning should be somewhat familiar with. Think twice before playing with them when you have no related experience. It's more art than science, meaning you need lots of time to experiment. Note I've already spent a week on docs... check out the updated Getting Started guide, in particular the Docker (Compose) installation instructions. We found those most important. Really need to finish multi user support now.

**Reply by rickysarraf** [reply]
Reverting [this commit](https://github.com/photoprism/photoprism/commit/e14209526586fe5115cc7a3e49aa76481b0a12bc) has very well served the purpose in making the behavior inline with what I had witnessed initially. Without this change, the overall detection of faces is far more better.

In my humble opinion, I would request the same that is [being said](https://github.com/photoprism/photoprism/issues/1587#issuecomment-950466822).

**Reply by lastzero** [reply]
see https://github.com/photoprism/photoprism/issues/1587#issuecomment-950753655

**Reply by ChristopherHaws** [reply]
I was trying to tweak these settings too and didn't really understand them, but from what I can tell they are pretty common terms/settings in machine learning/facial recognition so I just asked ChatGPT (we're all going to lose our jobs 😆😢😭) to provide me with some more info on these properties because I am not very literate in ML yet (been trying to learn more!). The responses were pretty decent so I am going to post them here in case it can help someone else out.

NOTE: I did not write these descriptions, ChatGPT did, so take these descriptions with a grain of salt. If someone wants to proof the accuracy of them and put them in the docs, it might help people trying to tweak these settings, but as the docs says, this is an advanced topic so 🤷 

Thanks again for this amazing piece of software! :)


# PHOTOPRISM_FACE_SIZE
Minimum size of faces in pixels refers to the smallest dimensions that a face can have in an image and still be detected and recognized by a facial recognition algorithm. This size is usually determined by the technical limitations of the algorithm and the desired level of accuracy. The minimum size of faces will depend on the specific algorithm being used, the image resolution, and the desired level of accuracy. A smaller size may result in lower detection accuracy, while a larger size may result in longer processing times or higher computational requirements.

# PHOTOPRISM_FACE_SCORE
Minimum face quality score refers to the minimum acceptable quality score assigned to a face by a facial recognition algorithm. This score reflects the confidence of the algorithm in the detection and recognition of the face. The score may depend on various factors such as the resolution, lighting, and pose of the face. The exact value of the minimum face quality score will depend on the specific algorithm being used and the desired level of accuracy for the application.

# PHOTOPRISM_FACE_OVERLAP
Face area overlap threshold refers to the minimum or maximum allowed overlap between two face bounding boxes for them to be considered as matching or belonging to the same cluster. The value of the threshold is set by the user and may depend on the specific facial recognition algorithm and application.

# PHOTOPRISM_FACE_CLUSTER_SIZE
The size of automatically clustered faces in pixels will vary depending on the specific facial recognition algorithms and implementation used. It can be influenced by factors such as the resolution of the images, the size of the faces in the images, and the desired level of accuracy. The size of the faces in pixels can be adjusted through image processing techniques such as resizing, scaling, or cropping. The size of automatically clustered faces may also depend on the minimum size threshold set by the user or determined by the algorithm.

# PHOTOPRISM_FACE_CLUSTER_SCORE
The minimum quality score of automatically clustered faces in facial recognition depends on the specific algorithm and implementation being used. The quality score is typically a metric that assesses the confidence of the facial recognition algorithm in the similarity or match between faces.

Some algorithms use a quality score that ranges from 0 to 1, where a score of 1 indicates a perfect match and a score of 0 indicates no match. The minimum quality score for a match to be considered as part of a cluster can be set by the user or designer of the system and can vary depending on the specific requirements and application.

For example, in a security system with high accuracy requirements, a higher minimum quality score may be set to ensure only highly confident matches are included in a cluster. On the other hand, in a system with lower accuracy requirements, a lower minimum quality score may be acceptable to allow for more relaxed matching criteria and potentially increase the number of faces forming a cluster.

In general, the minimum quality score for automatically clustered faces should be set based on a trade-off between accuracy and the desired number of faces forming a cluster.

# PHOTOPRISM_FACE_CLUSTER_CORE
The number of faces forming a cluster core in facial recognition is dependent on the specific clustering algorithm being used and the desired clustering behavior. Some algorithms may require a minimum number of faces to form a cluster, while others may allow for clusters with only a single face. The number of faces forming a cluster core also affects the robustness and stability of the cluster. A cluster with a larger number of faces is generally considered to be more robust and stable compared to a cluster with fewer faces. The optimal number of faces forming a cluster core is usually determined through experimentation and depends on the specific requirements and application of the facial recognition system.

# PHOTOPRISM_FACE_CLUSTER_DIST
A similarity distance threshold in facial recognition determines the maximum distance between faces that can still be considered as part of the same cluster. It is a measure of how similar or dissimilar two faces are based on the features extracted from their images. A low threshold will result in a more stringent cluster, where only highly similar faces are grouped together. A high threshold will result in a more relaxed cluster, where faces with a greater degree of difference are included. The optimal threshold is usually determined through experimentation, and it depends on the specific requirements and application of the facial recognition system.

# PHOTOPRISM_FACE_MATCH_DIST
A similarity offset in facial recognition refers to the minimum similarity score required for a face to be matched with an existing cluster. It acts as a threshold for face matching, defining the minimum distance between a face and the cluster centroid (representative face) that would allow the face to be considered a match. A higher similarity offset implies a stricter matching criterion, where only highly similar faces will be matched to existing clusters. On the other hand, a lower similarity offset allows for a more relaxed matching criterion, where faces with a lower similarity score can still be matched to existing clusters. The optimal similarity offset is usually determined through experimentation and depends on the specific requirements and application of the facial recognition system.

**Reply by ndfan77** [reply]
Some suggested values to create more aggressive face recognition behavior would be helpful.
